### ISSUE TITLE
Ractive doesn't always recognize object changes

### ISSUE THREAD

**Issue by matthew-dean** [ractivejs/ractive#1952]
One of my biggest challenges with Ractive has been that it doesn't seem to always recognize when the underlying data changes.

My use case: I have an app with a file tree. Building the file tree works great. However, I found that if I tried to attach a new tree, it would still render the old output. I essentially had to "destroy" the data and then attach new data to get it to re-render.

I found a suitable workaround, but now I have a new problem.

I have a "search" feature which filters the files and shows the number of "matches" next to the file.

Internally, I have a ```searchTree``` object and a ```directory``` object. When searching, I basically swap out the objects and attach them to the data root of the ractive view.

When search is cleared, I set the data back to ```directory```. BUT, here's where it gets weird. SOMETIMES (but not always), the Ractive will render the number of "matches" in the regular directory view when I switch back from search, even though they are from two completely data sources. They may list some of the same files in each dataset, but the file "objects" are different objects of the same prototype.

My template looks like:
```
<!-- {{>file}} -->
<li class="{{ type === 'folder' ? 'folder' : 'file' }} {{ ./open ? 'open' : '' }}">
	{{#if ~/root.search }}
	<h1 class="context-path">{{{ getPath(path, keywords) }}}</h1>
	{{/if}}
	{{#if ./selected }}<s class="selected"></s>{{/if}}
	<span on-click="highlight" on-dblclick="select"><i on-click="open"></i>{{#if 
		./matches && ./matches.length === 0 }}
		<s>{{{ highlightMatch(name, ext, keywords) }}}</s>
		{{else}}
		<s>{{ name }}<s class="{{ c ? 'c' : ''}} {{ f ? 'f' : '' }}">{{ ext }}</s></s>
		{{/if}}
	{{{ ./matches && ./matches.length ? ' <s class="m">('+./matches.length+')</s>' : '' }}}</span>
	{{# ./open }}
	{{# ./type === 'folder' }}
		{{>folder}}
	{{/ ./type === 'folder' }}
	{{ /open }}
</li>
<!-- {{/file}} -->
```
The ```matches``` property is set to zero for the ```directory``` file objects, yet, at times, it will render it regardless when the OTHER properties of the file match (name, extension, etc). I don't know how to tell Ractive "hey, these two objects are really different" or "please REALLY check that these objects and all of their properties are different".

Is there a way to dirty the whole set, or somehow tell Ractive to be more thorough in comparing objects before rendering? The weird thing is that these errors are not consistent. Usually, the two data sets render correctly, but, often, they show properties from each other's data. I know comparing objects in JS for equality can be tricky, that's why I'm wondering if I can just flag objects as dirty.

**Comment by Madgvox** [comment]
Since I don't have access to the source code, a couple of exploratory questions:

- How are you swapping out the data?

- If you're not doing it already, does calling `update()` on the ractive after you swap out the data set correctly swap out the data?

**Comment by martypdx** [comment]
> Is there a way to dirty the whole set, or somehow tell Ractive to be more thorough in comparing objects before rendering?

Ractive errs on the side of being overly aggressive in checking for object changes. If an object changes, all of the downstream dependents (properties and properties on down) are checked.

In my experience, in could be a) a bug, or  b) a reference bound to a different place in the data hierarchy than you expect, or c) something else. :)

> My use case: I have an app with a file tree. Building the file tree works great. However, I found that if I tried to attach a new tree, it would still render the old output. I essentially had to "destroy" the data and then attach new data to get it to re-render.

Can you post an example of the original issue with template and data? This newer ones seems more complex, and perhaps if we go back to this original issue it would be easier.

**Comment by matthew-dean** [comment]
**Update: Alright, after I wrote what's below, I just tried to get the second behavior to happen about a dozen times, and not once could I replicate it this time. So, I don't know if the issue is magically gone, or one of those things where you change something that doesn't trigger the bug, so I don't know. I saw it sporadically for a while and couldn't replicate it today. So you're welcome to not pay attention to anything below this point, but I've left it to just answer the question, and maybe you'll spot something else I did wrong.**

---

Sure, here's some code examples:
```
var currentDir = Files.currentDirectoryListing();
var ractive = new Ractive({
				el: document.getElementById('files')
				, template: filesTemplate
				, data: { 
					root: currentDir
			});

// in files module
F = Files = { currentDirectoryListing: function() {
			if(F.searchTree.length > 0 && F.searchTree[0].path === F.currentDir[0].path) {
				F.searchTree[0].open = true;
				F.searchTree[0].search = true;
				return F.searchTree[0];
			}
			if(F.currentDir.length === 0) {
				return F.currentDir;
			}
			else {
				F.currentDir[0].open = true;
				F.currentDir[0].search = false;
				F.currentDir[0].searchText = '';
				return F.currentDir[0];
			}
}
```
Without cutting / pasting too much code, basically: the currentDir property is "reset" when a new directory is selected. After that, I tried to do a regular ```update()```, and I could never get it to work. My solution was to rebind the root property:

```this.ractive.set('root', currentDir);```

So, that solved getting the whole tree to refresh. But it doesn't work when swapping the searchTree object for the currentDir object. It shows the right files, but once in a while, it "carries over" property values from the search tree into the file tree.

In other words, the search tree may have a file object somewhere in an array that looks like this:
```
{ 
	type: 'file', 
	name: 'myfile',
	ext: '.js', 
	path: '~/myfile.js', 
	matches: 3,
	c: true
	f: true
}
```
Now, I switch back to the directory listing and "re-bind" the ractive by setting the root property. The directory listing, somewhere, has an object that looks like this:
```
{ 
	type: 'file', 
	name: 'myfile',
	ext: '.js', 
	path: '~/myfile.js', 
	matches: 0,
	c: true
	f: true
}
```
However, once in a while (1 out of 10 times?), Ractive will render this file as:
```
  myfile.js (3)  // The 3 being the # of search matches in the template
```
In other words, it seems to merge object A with object B, and sets the value of object B ```matches``` to 3 instead of 0.

It's important to note that these two file objects are not the same objects. They have the same structure and are instances of the same prototypal function, but file objects are not "copied" from the search tree to the directory tree. They just happen to have the same file name / path.

What's curious is that the behavior is not consistent. If I switch back and forth a few times from search to the full listing, or repeat the same actions on app load, even though the data is the same, the template will show this "merging" behavior in different ways at different times. That is: sometimes it doesn't merge the ```matches``` property, and sometimes it does for some files / directories and not others.

**Comment by martypdx** [comment]
@matthew-dean does it possibly only happen when matches is `0`? I'm wondering if it's a falsey value bug.

**Comment by Madgvox** [comment]
@matthew-dean I'm going to assume this issue has been resolved and mark it closed. If you have any more questions / concerns, feel free to reopen the issue at any time.

**Comment by matthew-dean** [comment]
@Madgvox Unfortunately, I was never able to get this to stop happening, no matter what I tried. Ractive would render (some, but not all) previous values for (some) objects in a tree (some of the time, and different pieces of the tree each time, or none), no matter if I marked the entire object set as dirty. My workaround was to simply employ CSS to hide when Ractive would falsely render something.
